controllers/volumereplicationgroup_controller.go

```diff
@@ -146,6 +158,60 @@ func (r *VolumeReplicationGroupReconciler) configMapFun(configmap client.Object)
 	return req
 }
 
+func (r *VolumeReplicationGroupReconciler) waitForCrds() error {
+	log := r.Log.WithName("checkingDependencies")
+
+	requiredCrds := []string{volumeReplicationClass, volumeReplication}
+
+	for _, crd := range requiredCrds {
+		err := r.waitForResource(log, crd)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *VolumeReplicationGroupReconciler) waitForResource(logger logr.Logger, resourceName string) error {
+	unstructuredResource := &unstructured.UnstructuredList{}
+	unstructuredResource.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   volrep.GroupVersion.Group,
+		Kind:    resourceName,
+		Version: volrep.GroupVersion.Version,
+	})
+
+	err := r.Client.List(context.TODO(), unstructuredResource)
+	if err != nil {
+		if !meta.IsNoMatchError(err) {
+			logger.Error(err, "unexpected error occurred while checking CRD", "CRD", resourceName)
+
+			return err
+		}
```

Comment:
Code above can be extracted to a helper so we don't need to repeat it:

    func resourceExists(resourceName string) (ok bool, err error) {
        unstructuredResource := &unstructured.UnstructuredList{}
	    unstructuredResource.SetGroupVersionKind(schema.GroupVersionKind{
		    Group:   volrep.GroupVersion.Group,
		    Kind:    resourceName,
		    Version: volrep.GroupVersion.Version,
	    })

	    err := r.Client.List(context.TODO(), unstructuredResource)
	    if err != nil {
		    if !meta.IsNoMatchError(err) {
			    return err // let caller handle this
		    }
            return false, nil // Missing resource is not an error.
        }
        return true, nil // Exists  
    }

With this we can have simpler and more clear wait loop, and we don't need to repeat the boilerplate.

Reply:
But then we will repeat this code fragment with no reason:
unstructuredResource := &unstructured.UnstructuredList{} unstructuredResource.SetGroupVersionKind(schema.GroupVersionKind{ Group: volrep.GroupVersion.Group, Kind: resourceName, Version: volrep.GroupVersion.Version, })

Reply:
This is used only once inside resourceExists()? waitForCrds

Reply:
Done